src/functions.php

```diff
@@ -897,3 +897,38 @@ function _caseless_remove($keys, array $data)
 
     return $result;
 }
+
+/**
+ * UTF-8 aware \parse_url() replacement.
+ *
+ * The internal function produces broken output for non ASCII domain names (IDN)
+ * when used with locales other than "C".
+ *
+ * On the other hand, cURL understands IDN correctly only when UTF-8 locale is
+ * configured ("C.UTF-8", "en_US.UTF-8", etc.).
+ *
+ * @see https://bugs.php.net/bug.php?id=52923
+ * @see https://www.php.net/manual/en/function.parse-url.php#114817
+ * @see https://curl.haxx.se/libcurl/c/CURLOPT_URL.html#ENCODING
+ *
+ * @return array|false
+ * @internal
+ */
+function _parse_url($url, $component = -1)
```

Comment:
this should just be a private method on the Uri class to avoid exposing it.

Reply:
please also add type declarations

Reply:
Thanks, @Tobion, I moved it inside the Uri class. Please take a look at the new PR to `1.x` branch.